x/feeshare/ante/ante.go

```diff
@@ -0,0 +1,119 @@
+package ante
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	wasmtypes "github.com/CosmWasm/wasmd/x/wasm/types"
+	authforktypes "github.com/CosmosContracts/juno/v12/x/auth/types"
+	feeshare "github.com/CosmosContracts/juno/v12/x/feeshare/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/types"
+)
+
+// FeeSharePayoutDecorator Run his after we already deduct the fee from the account with
+// the ante.NewDeductFeeDecorator() decorator. We pull funds from the FeeCollector ModuleAccount
+type FeeSharePayoutDecorator struct {
+	ak             authforktypes.AccountKeeper
+	bankKeeper     authforktypes.BankKeeper
+	feegrantKeeper authforktypes.FeegrantKeeper
+	feesharekeeper authforktypes.FeeShareKeeper
+}
+
+func NewFeeSharePayoutDecorator(ak authforktypes.AccountKeeper, bk authforktypes.BankKeeper, fk authforktypes.FeegrantKeeper, fs authforktypes.FeeShareKeeper) FeeSharePayoutDecorator {
+	return FeeSharePayoutDecorator{
+		ak:             ak,
+		bankKeeper:     bk,
+		feegrantKeeper: fk,
+		feesharekeeper: fs,
+	}
+}
+
+func (fsd FeeSharePayoutDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return ctx, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
+	}
+
+	err = FeeSharePayout(ctx, fsd.bankKeeper, feeTx.GetFee(), fsd.feesharekeeper, tx.GetMsgs())
+	if err != nil {
+		return ctx, sdkerrors.Wrapf(sdkerrors.ErrInsufficientFunds, err.Error())
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+// FeePayLogic takes the total fees and splits them based on the governance params
+// and the number of contracts we are executing on.
+// This returns the amount of fees each contract developer should get.
+// tested in ante_test.go
+func FeePayLogic(fees sdk.Coins, govPercent sdk.Dec, numPairs int) sdk.Coins {
+	var splitFees sdk.Coins
+	for _, c := range fees {
+		rewardAmount := govPercent.MulInt(c.Amount).QuoInt64(int64(numPairs)).RoundInt()
+		if !rewardAmount.IsZero() {
+			splitFees = append(splitFees, sdk.NewCoin(c.Denom, rewardAmount))
```

Comment:
Should add check if c.Denom already have in splitFees. If so, should addition rewardAmount with the current amount of c.Denom in splitFees

Reply:
https://github.com/CosmosContracts/juno/pull/385/commits/69087122eabfdfb534edbe651207cc4078298418